### ISSUE TITLE
Service independence (Starting playback without opening app)

### ISSUE THREAD

**Issue by etyarews** [OxygenCobalt/Auxio#244]
#### Describe the feature you want to implement:
[While Auxio have Intents](https://github.com/OxygenCobalt/Auxio/issues/17), they are limited and don't work unless Auxio is already running in the background. My knowledge of programming is limited, but it appears the latter is caused by the type of intent.

#### Is your feature request related to a problem? Please describe:
I'm trying to control music playback through Auxio, namely I'm a user of Tasker and sometime between Android 7 and Android 12 it became unreliable to start music playback by just simulating a Media Button press. Intents appear to be the only solution left, and sadly the ones Auxio have are of the broadcast type. 

#### Do other music players handle this? If so, how?
PowerApp actually does have it, they are of the Service type and I'm able to start them even after a reboot, and it has a long list of possible intents that I'm aware of:

- Play/Pause
- Pause
- Resume
- Stop
- Next
- Previous
- Start Fast Forward
- Stop Fast Forward
- Start Fast Rewind
- Stop Fast Rewind
- Toggle Repeat
- Toggle Shuffle

#### Why do you think this will improve everyone's usage of Auxio?
While not useful to the majority of Auxio users, it will be incredibly helpful to the ones that are also Tasker users, and because it isn't a request for a Tasker Plug-in, it has could help users of that can send Intents.

#### Due Diligence:
- [x] I have read the [Contribution Guidelines](https://github.com/OxygenCobalt/Auxio/blob/dev/.github/CONTRIBUTING.md).
- [x] I have read the [Accepted Additions and Requests](https://github.com/OxygenCobalt/Auxio/blob/dev/info/ADDITIONS.md) document.
- [x] I have checked for this feature in the [FAQ](https://github.com/OxygenCobalt/Auxio/blob/dev/info/FAQ.md).

**Comment by OxygenCobalt** [comment]
This is not possible. 

To make my app work in modern android, I am effectively forbidden from having any means of interaction that does not require you to open the app first.

Android 8 made foreground services (Which includes Auxio's music playback) MUCH harder to initialize. I have 5 seconds to initialize everything or the app will crash. Factoring in music loading and the restoration of the playback state, I simply cannot fit it into that window.  

The options I have to work around it are:
- Create a dummy notification that allows me to initialize immediately, but then I run into issues with rate-limiting and display coherency. It would cause more issues than it solves.
- Perpetually trap my app in "API 2X hell" like Poweramp and Phonograph so I can skirt the restrictions within the compatibility sandbox.

And then I would have to work Android 12's insane restriction on launching foreground services when the app is backgrounded. If I were to add these intents, it could open up a footgun where someone on Android 12 tries to fire an intent while Auxio is backgrounded, and then gets confused why the app crashes.

It's just not practical anymore.

The days of being able to do neat things without interaction on Android are over. I do not want my app to be an awful, buggy mess, so I require you to open Auxio before doing anything. If this means I have to annoy a vanishingly small minority of Tasker users, so be it.

**Comment by etyarews** [comment]
Hey if you don't mind just a side note, so I don't have to open another issue for something that can be easily answered:

What about _Static_ Shortcuts? Auxio has a Dynamic one that doesn't show up both in Tasker and Kustom, and other non-launcher apps that can initiate shortcuts. I've managed to _in a really dumb roundabout way_ trigger the Dynamic one with Tasker, but it would be far easier and more flexible if it was possible for you to implement a couple Static ones(or even just one). Again, this would also benefit Kustom users, and others.

I also _think_ Tasker is already using the `MediaController` API, as we can get Media information and interact as if it was media buttons as I stated. But even if I'm wrong, the google documentation mentions its only for ongoing media session, I don't think it is possible to start a media session from nowhere without initializing the app first.

**Comment by OxygenCobalt** [comment]
> What about Static Shortcuts? Auxio has a Dynamic one that doesn't show up both in Tasker and Kustom, and other non-launcher apps that can initiate shortcuts. I've managed to in a really dumb roundabout way trigger the Dynamic one with Tasker, but it would be far easier and more flexible if it was possible for you to implement a couple Static ones(or even just one). Again, this would also benefit Kustom users, and others.

I use dynamic shortcuts so that shortcuts in debug builds don't redirect to release builds or just not work at all. Maybe I can do some cursed hack where the release shortcut is static and the debug shortcut is dynamic.

> I also think Tasker is already using the MediaController API, as we can get Media information and interact as if it was media buttons as I stated. But even if I'm wrong, the google documentation mentions its only for ongoing media session, I don't think it is possible to start a media session from nowhere without initializing the app first.

Yeah, I already know. The issue is that the addition of service Intents would open the door for the app  to be initialized without opening (Which is impossible), while `MediaController` doesn't allow that.

**Comment by OxygenCobalt** [comment]
Okay, let me clarify that this might not be *entirely* impossible. It's just extremely difficult and not something that I actively want to say I have planned. 

Over time I tend to warm up to features as I think on and off about them and eventually develop a system that will solve the problem sensibly. If I'm confident that the odds have shifted and the addition is a net benefit with the new approach, I create a new issue with my proposal and signal that it's actively planned and not just something I'm thinking about.

**Comment by etyarews** [comment]
Thank for your consideration, but just to clarify: are you talking about intents or static shortcuts?

**Comment by OxygenCobalt** [comment]
Intents. The static shortcuts thing I need to do more research on but should be slightly more possible.

**Comment by OxygenCobalt** [comment]
Okay, turns out your feature request might be required for #37 because the `MediaBrowserService` lifecycle is absurd. I haven't looked into it very much though.

**Comment by etyarews** [comment]
So it's going to get open again?

I will not lie, I don't totally understand the technical implications or even why it is required for Android Auto, but hey, it is still an improvement

**Comment by OxygenCobalt** [comment]
Maybe. The [documentation](https://developer.android.com/guide/topics/media-apps/audio-app/building-a-mediabrowserservice) is extremely vague, but it implies that Android Auto, can start the playback service at any time, including when the app is completely dead. This is the same avenue that you want with your intents, except this is required for me.

**Comment by etyarews** [comment]
I never even _looked_ at Android Auto, but thinking about it, it doesn't make sense to have to manually open a Music Player while in a car.

Best case would be to test and see if a Google app can auto start with Android Auto.... except last I checked, Google doesn't make any music player anymore

**Comment by OxygenCobalt** [comment]
Youtube Music exists. I'm going to add the media browser boilerplate and see what happens in an emulator. 

Assuming that I need to handle this case, I think I have a route to implementing it. Basically, I would need to unify the two services in Auxio (music loading and playback) into a single service that can load and play music. The implementation details are hard and annoying and blocked by countless things, so this is still low-priority.

**Comment by etyarews** [comment]
If you don't mind me asking, why is Auxio so _aggressive_ with music loading? I don't think I've ever seen an app being so aggressive and upfront about checking for changes to the library.

1. It reloads it every time you start the app, even if it was seconds since the last time it checked it
2. It doesn't happen in the background, meaning it stops the user from using the app.
3. "Ignore MediaStore Tags" takes significant time to reload (which is fine), but because of the above points it isn't really a viable option.

**Comment by OxygenCobalt** [comment]
A few reasons, I guess:
1. I used to use Music Player GO, and so a bunch of Auxio's architecture is derived from that app. MPGO is similarly aggressive with loading.
2. Being aggressive with loading ensures a more sensible music experience at the cost of lost time at the start. I've seen how background loading mechanisms work in other apps (Read: Vanilla) and they are buggy garbage that usually renders the library similarly unusable until it's done loading.
3. I'm not good at programming clever state management systems like what I assume Poweramp does to implement seamless background loading.

Note that the next version will have a much faster music loader, as I am now caching song metadata.

**Comment by etyarews** [comment]
Eeeh, I'm going to have to look at the next version first to give it some thought and see what really changed in terms of the user experience, because I'm not sure about the implications of caching song metadata without seeing it in action.

But I don't really get why it isn't doing like VLC by doing a big scan on boot and on the background, an option to manually refresh the library on the 3-dot menu would also help. The probability of the user changing the music library (specially if they are using a whitelist of folders) and not remembering it by doing a manual re-scan is rather low. Furthermore, you could mitigate the number of auto-refreshes by comparing when was the last time it was done, as it redo it even if you closed and reopened the app in less than 5 secs, but I'm assuming this will be taken care of with the caching.

**Comment by OxygenCobalt** [comment]
The caching implementation is mostly to circumvent the time sink within "Ignore MediaStore tags", that being the quite slow file opening/metadata extracting process. It works something like this:

1. Query the media database for every song on the device
2. For each song, check with the cache if it has extracted metadata for that song that can be used. This is used by comparing the timestamps of the song from the media database and the timestamps in the cache.
3. If we can use the cached metadata, we are done.
4. Otherwise, we run the slower metadata extraction routine and save that song to the cache.

> But I don't really get why it isn't doing like VLC by doing a big scan on boot and on the background

AFAIK, newer android versions will not allow me to wake my app on boot anymore unless the user disables battery optimizations. I really want the app to work out of the box without having to make the user fiddle with hidden android settings, so I can't implement this.

> The probability of the user changing the music library (specially if they are using a whitelist of folders) and not remembering it by doing a manual re-scan is rather low. 

Not really sure what you mean by this.

> Furthermore, you could mitigate the number of auto-refreshes by comparing when was the last time it was done, as it redo it even if you closed and reopened the app in less than 5 secs, but I'm assuming this will be taken care of with the caching.

That's basically the cache mechanism, albeit done on a song-by-song basis. You also have to consider that I still have to load the cache as well, hence why I said there will always be some overhead.

**Comment by etyarews** [comment]
> The caching implementation is mostly to circumvent the time sink within "Ignore MediaStore tags", that being the quite slow file opening/metadata extracting process. It works something like this:
> 
>     1. Query the media database for every song on the device.
> 
>     2. For each song, check with the cache if it has extracted metadata for that song that can be used. This is used by comparing the timestamps of the song from the media database and the timestamps in the cache.
> 
>     3. If we can use the cached metadata (the timestamps are the same, so the song has not changed), we are done.
> 
>     4. Otherwise, we run the slower metadata extraction routine and save that song to the cache.
> 
> 
> So, if you don't change your library, loads should go from taking ~15 seconds for a 1k song library to only 1 second or lower. If you only change part of your library, then it should still be quite fast, with a bit more time being needed to scan the new music.

Yeah, the caching part seems fair. However, if it is _really_ 1 second or lower, you might want to consider "hiding" it behind a fake app loading screen that takes half to one second to complete and the rest of the loading is shown as usual. This way, the app will appear faster than it actually is.

> > But I don't really get why it isn't doing like VLC by doing a big scan on boot and on the background
> 
> AFAIK, newer android versions will not allow me to wake my app on boot anymore unless the user disables battery optimizations. I really want the app to work out of the box without having to make the user fiddle with hidden android settings, so I can't implement this.

Alright, so I'm on Android 12, but there are 2 things which _might_ minimize your concerns:

1. Tasker can show a system floating dialog that directly asks you whether you want to disable battery optimizations.
2. Kustom apps show a floating dialog that concisely tells the user how to disable battery optimization, and it takes the user directly to the android setting screen, so it's not hidden.
3. Both apps somehow can know if the battery optimization was disabled or not, re-enabling it makes those apps ask again for the user to disable it. 

The third point is important, because you can use it as a means to fall back to the previous method. Or make the on boot an option in the settings that only nags the user when they open the app for the first time.
 
> > The probability of the user changing the music library (specially if they are using a whitelist of folders) and not remembering it by doing a manual re-scan is rather low.
> 
> Not really sure what you mean by this.
 
AFAIK Auxio works in two "modes":

- Whitelist: only look for things on specific folders
- Blacklist: Look for things everywhere, except specific folders.

If the user is using the "Whitelist" method, they are more likely to be aware of changes they made to those folders, meaning they are capable of manually triggering a re-scan, _if_ there was an option on the 3-dot menu AND scan _only_ happened on boot.

Basically, "Whitelist-users" would be fine if scan only happened on boot.

**Comment by OxygenCobalt** [comment]
> Yeah, the caching part seems fair. However, if it is really 1 second or lower, you might want to consider "hiding" it behind a fake app loading screen that takes half to one second to complete and the rest of the loading is shown as usual. This way, the app will appear faster than it actually is.

Let me clarify that it's 1 second or so for an "average" library (1000 songs), larger libraries will slow the loading process down more.

> Alright, so I'm on Android 12, but there are 2 things which might minimize your concerns:

But that's the thing: I don't want to have to make an annoying onboarding flow just to disable  battery optimizations. I want my app to work without much fiddling by the user. Disabling the optimizations is the "easy way out" that will only likely get harder and harder in future versions.

**Comment by etyarews** [comment]
> Let me clarify that it's 1 second or so for an "average" library (1000 songs), larger libraries will slow the loading process down more. Because of this, I may as well keep the loading UI as-is.

I know, that's not what I mean.

Assuming:
1. Cached loading will take between 0.5 to 1 second to load a library with 1000 songs
2. You know with certainty that the vast majority of users don't keep more than 1000 songs on their devices.

You can then make a fake full screen splash screen on top of the loading dialog. It is dismissed if:
- The cached loading has finished
OR 
- It started loading things that are not in cache
OR
- One second has passed

In this case, most users wouldn't even know Auxio is loading the library. The ones that have bigger libraries will have Auxio open into a loading dialog that is already in the halfway point, meaning that for them, it will also appear like Auxio is faster than it actually is. 

Granted this adds complexity, and I'm not even 100% into the idea, but eeh, I needed to explain a little since Loading Screens are really annoying for end users, specially ones that appears every time you cold start the app and don't allow you to interact with the app while it is loading.

>Besides, disabling the optimizations is the "easy way out" that will only likely get harder and harder in future versions.

I understand your point, and can't really deny them, however, I'm not 100% sure about this point in specific. Android 13 added a [convenient way to check for active apps in the notification drawer](https://developer.android.com/static/images/guide/components/fgs-manager.svg), this seems to point to me that Google is considering backtracking a bit on the battery limitations as long as they find an easy way to communicate it to the user

**Comment by OxygenCobalt** [comment]
> I know, that's not what I mean
> ...

I see. Doing that is actually [not recommended by google anymore](https://developer.android.com/develop/ui/views/launch/splash-screen/migrate#remove_the_custom_splash_screen_activity). The splash screen is only intended for basic app initialization. Otherwise for loading I should use placeholders or indicators, like I currently am. If I were to shove in my own splash to hide loading, that would just create needless jank that I don't want.

> Besides, disabling the optimizations is the "easy way out" that will only likely get harder and harder in future versions.

The FGS is merely another way to cripple apps under the guide of "battery protection". You now have the free ability to completely kill an app (not just the service, the WHOLE APP) if it has an active foreground service, making services even harder to manage. I don't even have the leeway to save the playback state or reset the widget when it's pressed, my app is just killed.

**Comment by OxygenCobalt** [comment]
Okay, I'm going to re-open this. I realized that if I unify my two services, that more or less resolves the issues of background starts and the foreground time limit. It's hard, but doable, and it will actually enable #37. I think I'll try to do it alongside #257 in a more general refactor of the service lifecycle.

**Comment by OxygenCobalt** [comment]
Actually, may not necessarily need to unify the music loading and playback services. Seems like the type of work the music loading service is doing will be deprecated in future versions. To futureproof it, I'll have to use the new "work manager" system. No clue how much easier or harder this becomes.

**Comment by etyarews** [comment]
A bit of off-topic, but since this conversation started because of it, could you consider adding a Tasker Plugin?

The Tasker dev made a video explaining how to do it: https://www.youtube.com/watch?v=48IVJgDtu6Y

And the documentation seems... _relatively_ easy enough

**Comment by OxygenCobalt** [comment]
I can, but it's still blocked by this issue since Auxio physically cannot work correctly with Tasker without this issue.

**Comment by etyarews** [comment]
Don't the shortcut implies that Auxio already has the meaning to work if an app manually calls it?

As I said a while ago, what prevents me from using Auxio in Tasker is because it uses dynamic shortcuts, even then, I managed to use a truly cursed method (a hell to set-up and configure) to make Trigger the Shuff Dynamic Shortcut and it worked.

**Comment by OxygenCobalt** [comment]
Not quite. The shortcut opens the app's activity, which then delivers commands to the services. What I'm talking about is the ability to cut the app activity out, and simply deliver the command to the service directly. Auxio doesn't really support this.

Oh, also scrapping the "work manager" thing I brought up earlier. Turns out it's too OEM-dependent and once again runs into the same foreground start limitation insanity as prior.

**Comment by etyarews** [comment]
Would it be possible to just open a dialog or something and immediately close it?

**Comment by OxygenCobalt** [comment]
Too janky of a solution for something that can be solved in other ways.

**Comment by etyarews** [comment]
I mean, everytime we open Auxio, it opens a dialog about updating the library anyway.

Even if you found a perfect solution that can start playback from cold boot, it would still need to load the music library, no? By that point you would need to create a notification or a dialog anyway

**Comment by OxygenCobalt** [comment]
Oh, that's what you mean. Yes, I would have to show a notification. The issue is that the service performing playback and the service loading music are separate right now. Arguably, I could make the playback service show a fake notification to work around the limitation as it waits for music to load, but I could also just merge the two services and solve the problem that way.

**Comment by OxygenCobalt** [comment]
Regarding our discussion from awhile back, #384 is actually likely going to make it that Auxio will actually stop indicating that it's loading music front-and-center in some cases. It's still in flux however which cases Auxio will no longer indicate it, however.

**Comment by etyarews** [comment]
It's been 84 years. Any update on this?

Just to refresh, what I actually want is either Tasker Plugin implementation, or, a static shortcut that can play music through Auxio.

**Comment by OxygenCobalt** [comment]
Okay @etyarews, so I haven't been able to do much regarding this is because of:

1. School
2. Playlists
3. More School

You can become a [sponsor](https://github.com/sponsors/OxygenCobalt) though and prioritize this issue, making me work on it first before anything else (No guarantees on me implementing it by some date though, again because school). An $8/month sponsor has already done that with #342, so I'm doing that first unless a $16/month sponsor prioritizes something else. After that, it'll either be #322 or this issue, depending on which one I want to do.

**Comment by OxygenCobalt** [comment]
This has been prioritized by @gtsiam, @etyarews. I will be implementing it after #342. If you also prioritize it, I'll do it before #342.

**Comment by etyarews** [comment]
I'm sponsoring it for now. Can't do more cause exchange rates aren't really helping me right now. Also have no idea how to prioritize it officially, will figure it out later

**Comment by OxygenCobalt** [comment]
Yep, this issue now has $12 backing it, so it will be done first @etyarews.

**Comment by OxygenCobalt** [comment]
I'm making slow, incremental progress on making this work. The playback and music services are now combined, but now I have to operate under the assumption that I can never briefly exit foreground whatsoever. This is really hard and I think I need to plan it out with a flow chart...

Question: How are you going to send commands to Auxio @etyarews? I don't want to pay for Tasker to figure this out. Is it a broadcast, a startService call, or direct operation of the media session?

**Comment by etyarews** [comment]
Alright, so here's the documentation about making a Tasker Plugin:

https://tasker.joaoapps.com/pluginslibrary.html

There's roughly three concepts:

1. **Action**: Tasker will do something to the app. I.e. Tasker will trigger Auxio to play, either the current playlist, or a song, or another playlist.
2. **Event Condition**: Something happened in the app. I.e. Auxio started playing
3. **State Condition**: Something is/isn't happening to the app. I.e. Auxio is _playing_(which implies Tasker can know when it is not)

I imagine static shortcuts are reasonably similar to what an Action could be in Auxio, so you might want want to start with that

**Comment by OxygenCobalt** [comment]
Thanks for that @etyarews. I'm still not fully sure how to route the intents to PlaybackService. I think I'll try to use the MediaBrowserService intents.

Anyway here's a progress update:

So I've merged the two services, kind-of. What used to be the services now inherit a weird "fragment" class I made up. But they are still operating under the assumption that the app is in the foreground. The issue is that the flow might be something like:

- Music starts loading, so the service goes foreground
- Music finishes loading, so the service exits foreground
- <200ms of delay as Playback initializes
- Playback goes foreground
- Android 12+ gets angry and kills our app because gosh forbid you go background in any capacity (It's okay if you use firebase though. Because google cloud needs to hit it's quarterly revenue targets.)

And that's if there's music to be loaded and things to play. If there's neither, no foreground will occur and eventually the app is killed by the system for not starting foreground in time. As a result, I must design the service lifecycle under two hard rules:

1. Absolutely no downtime. Even if you're not doing work, hang on the last foreground notification until you're absolutely certain we can safely exit foreground without suprises. I think I can do this by basically making the service "fragments" flag if they want to foreground, background, or if it doesn't know what it wants to do yet.
2. Always assume you have basically no time to go foreground. OEMs probably play fast and loose with the foreground time limit, so I may have to start foreground with a fake notification just to satisfy the OS.

There's likely more refinement I need to do regarding this. I have some ideas on how to detect from the service's end if it was started by another part of the OS or if it was started by the app, which should allow me to switch between uptight and loose foreground behavior. Sadly I'm tied up with all the 3.4.0 bug reports, school, and a hackathon, so I'm not going to be able to do much for now.

**Comment by etyarews** [comment]
>And that's if there's music to be loaded and things to play. If there's neither, no foreground will occur and eventually the app is killed by the system for not starting foreground in time

So, you can start the service in the background, the issue is to keep it in the background?

**Comment by OxygenCobalt** [comment]
- The app is opened and starts the service
- Playing something causes the service to go into a foreground state, which is allowed since the *app is in the foreground* (This is the assumption the services operate on)
- Removing Auxio from recents (assuming it's still playing) leaves the service in foreground, but now the app is in the background
- Stopping playback then causes the service to go background, however in that case the system will immediately kill the app since it has no reason to use resources anymore

@etyarews

**Comment by OxygenCobalt** [comment]
The only current circumstance in the app where the service feasibly goes from foreground -> background -> foreground is when playback is stopped while the activity is in recents, and then a media button event starts playback again. I think that's an exception though since it's a broadcast. I might have to confirm that though or otherwise my entire idea of how to go about this change is flawed.

**Comment by etyarews** [comment]
Alright, so I've glanced around android docs and it appears that a foreground service needs to start a notification or something in order to count as "foreground" is that right?

**Comment by OxygenCobalt** [comment]
Yeah, going foreground requires a notification to indicate such.

**Comment by etyarews** [comment]
Yeah, but isn't this solved?

1. start foreground service
2. Create notification showing the library is loading
3. Start playback

**Comment by OxygenCobalt** [comment]
Number 2/3 assumes that music hasn't already been loaded and playback hasn't occurred, and does not handle the possible gap that occurs in foreground state between 2 and 3. @etyarews

**Comment by etyarews** [comment]
Isn't the point to merge the two states into one that can handle both?

**Comment by etyarews** [comment]
If it is the same service, wouldn't there be no actual gap for Android to see? Like, I understand there is a gap to execute it, but it is the same service just waiting for code to finish, no?

**Comment by etyarews** [comment]
As long as there is a notification it should be fine

**Comment by OxygenCobalt** [comment]
Yes, and that's the change I have to make where I must hang on an existing notification when I exit foreground @etyarews. That's somewhat difficult since I need to know when some code is not certain that it will go foreground, which in turn requires it to indicate that it won't go foreground.

**Comment by etyarews** [comment]
Why exactly do you need to exit foreground? You could start playback as a foreground service and _then_ dismiss the notification

**Comment by OxygenCobalt** [comment]
You can't dismiss foreground notifications. You can only enter and exit foreground.

**Comment by etyarews** [comment]
Does media playback controls count as a notification?

**Comment by etyarews** [comment]
But if it counts as a notification then it doesn't get killed for being a foreground service without foreground

**Comment by OxygenCobalt** [comment]
Basically:
- Service explicitly started as foreground -> Foreground allowed
- Service started while app is visible -> Foreground allowed
- Service stops foreground and then goes foreground later -> I don't know if this is allowed

**Comment by etyarews** [comment]
My confusion is because I don't see why it needs to stop being foreground. Can a foreground service starts another foreground service?

**Comment by OxygenCobalt** [comment]
> Can a foreground service start another foreground service

A foreground service? Yes. A background service? No. To not unify would mean that I would have to create a fake notification no matter what just to have the ability to spin up the music service, which I severely dislike.

**Comment by etyarews** [comment]
Can't you just:

1. Start Foreground Service to deal with library.
2. Create notification to not kill the service
3. Check if the library is already loaded, if not, then load it
4. Anyway, when it ends, start Foreground Service to use Playback 
5. Create media notification for the Playback Foreground Service.
6. Finish the Library Loading Service and dismiss its notification

**Comment by OxygenCobalt** [comment]
Number 2 is the deal breaker. I do not want a fake notification, I know other music players have pulled it off.

**Comment by etyarews** [comment]
Alternatively, if the Media Notification counts as a notification, you can just use that during loading, no? You could remove some buttons and change title to say something like "Now Loading"

**Comment by OxygenCobalt** [comment]
You might be right about this. I'm going to perform an experiment to see if this background restriction only applies once (i.e once I start foreground, I then have the free ability to go background and foreground whenever I want).

**Comment by OxygenCobalt** [comment]
Okay, this issue has spun out of control and has turned into a full re-architecture around Media3. If I mitigate the performance issues and figure out if there's a way to integrate space for the long-running music loading task, Media3 should automate most of this absurd notification lifecycle stuff away.

**Comment by etyarews** [comment]
This is getting way out of hand, please make sure if the change to Media3 will allow cold start before starting work on this

**Comment by OxygenCobalt** [comment]
Yep, this absolutely should @etyarews. I'm primarily doing this because it seemingly automates a lot of the service lifecycle management and should allow cold starts.

**Comment by OxygenCobalt** [comment]
Okay, here's the first demo of Auxio with an independent service @etyarews:

[Auxio_Canary.zip](https://github.com/OxygenCobalt/Auxio/files/14963237/Auxio_Canary.zip)

- I'm going to corner cut and say that if you don't make Auxio to do *something* (i.e play) after starting the service independently, you should expect it to crash. I seriously doubt anyone wants to intentionally start Auxio and then just do nothing. I might change this expectation eventually.
- No tasker plugin, still trying to perfect the Media API and Android Auto support
- No custom command set, just use MediaController since that's the standard
- Shortcut is still dynamic (I need to work on this)

I'm sorry this took so long. If you can do certain things now, please test them and get back to me. If you need a tasker plugin, I'll need more time.

**Comment by etyarews** [comment]
Thanks man, really appreciate it.

That said, I'm rather confused how I might be able to control this debug version using Tasker, I tried using Media Control, but it doesn't seem to work, just like the normal Auxio version.

Will have to test a bit and see if I'm doing something wrong, if Tasker is doing something wrong, or if it is an issue with Auxio.

**Comment by OxygenCobalt** [comment]
Okay, I actually took the time to test with tasker, and yeah it just doesn't work since it forwards to a BroadcastReceiver. I have to make a custom command set or a tasker plugin, basically. I'll have to work on that @etyarews.

**Comment by OxygenCobalt** [comment]
Okay, so @etyarews you could theoretically start a service with a forged MediaButtonIntent. This could possibly work, but the issue is that it would likely come before the service could fully initialize and would functionally be a no-op. I think I'll have to implement my own handling so I am able to tell when I'm able to forward playback actions and when I can't. The service can technically stand on it's own two feet though, so I'm closing it and spinning off the command handling into another issue.

**Comment by etyarews** [comment]
So, how would you run this service? Tasker can execute shell, and I think it can start with am startservice blablabla

**Comment by OxygenCobalt** [comment]
I believe you can make a System -> Send Intent action and then specify the intent name, package name, action, to line up with a `MediaButtonIntent`. This is really hard and fiddly, so while you could mess with that in the mean time I'm going to work on a tasker plugin I think. Tasker's default media controls are horrible @etyarews

**Comment by etyarews** [comment]
I'm just a little confused cause with the wording you used, it appears that Auxio can now start playing music without running in the background. So this part is done.

But you didn't mention what service needs to be called or how to do it, even ignoring Tasker as whole, so I have no way to test if it even works, which is making me rather confused. Like, how do you even start this service at all, is it possible to start with shell?

**Comment by OxygenCobalt** [comment]
> I'm just a little confused due to the wording used, as it implies that Auxio can now start playing music without running in the background, as an independent service was implemented.

This is true in theory. In practice the workflow is currently just
- Start the service (probably with a flag telling it that it needs to be a foreground service).
- Delay for an arbitrary period (5 seconds will do if your music library isn't too big)
- Start sending media control actions

> But you didn't mention what service needs to be called or how to do it, let's ignore Tasker for a sec here, like, how are you starting this independent service?

Auxio is starting it as it did before, more or less. It doesn't send any commands through intents still, the backend infra is still being used as before.

To clarify: I reconfigured this issue to focus on a small part of what you wanted. In general, you want tasker/automation support, which requires custom command handling infrastructure, which requires service independence (this issue). Sorry for not being clear here @etyarews
